src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -107,32 +117,278 @@ internal override IEnumerable<TSource> GetEnumerable(int index)
+                // Walk back through the chain of ConcatNEnumerableIterators looking for the one
                 // that has its _nextIndex equal to index.  If we don't find one, then it
                 // must be prior to any of them, so call GetEnumerable on the previous
-                // Concat2Iterator.  This avoids a deep recursive call chain.
-                ConcatNIterator<TSource> current = this;
+                // Concat2EnumerableIterator.  This avoids a deep recursive call chain.
+                ConcatNEnumerableIterator<TSource> current = this;
                 while (true)
                 {
                     if (index == current._nextIndex)
                     {
                         return current._next;
                     }
 
-                    ConcatNIterator<TSource> prevN = current._previousConcat as ConcatNIterator<TSource>;
+                    var prevN = current._previousConcat as ConcatNEnumerableIterator<TSource>;
                     if (prevN != null)
                     {
                         current = prevN;
                         continue;
                     }
 
-                    Debug.Assert(current._previousConcat is Concat2Iterator<TSource>);
+                    ConcatIterator<TSource> prev2 = current._previousConcat;
+                    Debug.Assert(prev2 is Concat2EnumerableIterator<TSource> || prev2 is Concat2CollectionIterator<TSource>);
                     Debug.Assert(index == 0 || index == 1);
-                    return current._previousConcat.GetEnumerable(index);
+                    return prev2.GetEnumerable(index);
                 }
             }
         }
 
+        private sealed class Concat2CollectionIterator<TSource> : ConcatCollectionIterator<TSource>
+        {
+            private readonly ICollection<TSource> _first;
+            private readonly ICollection<TSource> _second;
+
+            internal Concat2CollectionIterator(ICollection<TSource> first, ICollection<TSource> second)
+            {
+                Debug.Assert(first != null && second != null);
+                _first = first;
+                _second = second;
+            }
+
+            internal override int Count => checked(_first.Count + _second.Count);
+
+            public override Iterator<TSource> Clone()
+            {
+                return new Concat2CollectionIterator<TSource>(_first, _second);
+            }
+
+            internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
+            {
+                var nextCol = next as ICollection<TSource>;
+                if (nextCol != null)
+                {
+                    return new ConcatNCollectionIterator<TSource>(this, nextCol, 2);
+                }
+                return new ConcatNEnumerableIterator<TSource>(this, next, 2);
+            }
+
+            internal override void CopyToUntil(TSource[] array, int lastIndex)
```

Comment:
> Since we only hold a reference to the last collection that was concat'd, if we had to implement CopyTo then we would have to backtrack all the way to the least-recently concat'd collection to ensure we were copying to the correct index

Why?  We know how long the collection, and the first thing the implementation does is subtract that amount from the lastIndex.... why can't the caller do that instead?